client/commands/query/get.go

```diff
@@ -51,65 +52,93 @@ func Get(key []byte, prove bool) (data.Bytes, uint64, error) {
 		resp, err := node.ABCIQuery("/key", key, false)
 		return data.Bytes(resp.Value), resp.Height, err
 	}
-	val, h, _, err := GetWithProof(key)
+	val, h, _, _, err := GetWithProof(key)
 	return val, h, err
 }
 
 // GetWithProof returns the values stored under a given key at the named
 // height as in Get.  Additionally, it will return a validated merkle
 // proof for the key-value pair if it exists, and all checks pass.
-func GetWithProof(key []byte) (data.Bytes, uint64, *iavl.KeyExistsProof, error) {
+func GetWithProof(key []byte) (data.Bytes, uint64,
+	*iavl.KeyExistsProof, *iavl.KeyNotExistsProof, error) {
+
 	node := commands.GetNode()
+	cert, err := commands.GetCertifier()
+	if err != nil {
+		return nil, 0, nil, nil, err
+	}
+	return getWithProof(key, node, cert)
+}
+
+func getWithProof(key []byte, node client.Client, cert certifiers.Certifier) (
+	val data.Bytes, height uint64, eproof *iavl.KeyExistsProof, neproof *iavl.KeyNotExistsProof, err error) {
 
 	resp, err := node.ABCIQuery("/key", key, true)
 	if err != nil {
-		return nil, 0, nil, err
+		return
 	}
-	ph := int(resp.Height)
 
 	// make sure the proof is the proper height
 	if !resp.Code.IsOK() {
-		return nil, 0, nil, errors.Errorf("Query error %d: %s", resp.Code, resp.Code.String())
-	}
-	// TODO: Handle null proofs
-	if len(resp.Key) == 0 || len(resp.Value) == 0 || len(resp.Proof) == 0 {
-		return nil, 0, nil, lc.ErrNoData()
-	}
-	if ph != 0 && ph != int(resp.Height) {
-		return nil, 0, nil, lc.ErrHeightMismatch(ph, int(resp.Height))
+		err = errors.Errorf("Query error %d: %s", resp.Code, resp.Code.String())
+		return
 	}
-
-	check, err := GetCertifiedCheckpoint(ph)
-	if err != nil {
-		return nil, 0, nil, err
+	if len(resp.Key) == 0 || len(resp.Proof) == 0 {
+		err = lc.ErrNoData()
+		return
 	}
-
-	proof := new(iavl.KeyExistsProof)
-	err = wire.ReadBinaryBytes(resp.Proof, &proof)
-	if err != nil {
-		return nil, 0, nil, err
+	if resp.Height == 0 {
+		err = errors.New("Height returned is zero")
+		return
 	}
 
-	// validate the proof against the certified header to ensure data integrity
-	err = proof.Verify(resp.Key, resp.Value, check.Header.AppHash)
+	check, err := getCertifiedCheckpoint(int(resp.Height), node, cert)
 	if err != nil {
-		return nil, 0, nil, err
+		return
 	}
 
-	return data.Bytes(resp.Value), resp.Height, proof, nil
+	if len(resp.Value) > 0 {
+		// The key was found, construct a proof of existence.
+		eproof = new(iavl.KeyExistsProof)
+		err = wire.ReadBinaryBytes(resp.Proof, &eproof)
```

Comment:
Spoke to @ebuchman about this - we could indeed pass `nil` for the value, but it would only work for these two proofs. Will look into this separately, as it's a change to merkleeyes.